File: legacy/main/rs/emulate/legacy/version/CrcList.java
package rs.emulate.legacy.version;

import com.google.common.base.Preconditions;

/**
 * A list of file CRCs, for a particular ArchiveEntry (such as models).
 * <p>
 * This class is immutable.
 *
 * @author Major
 */
public final class CrcList {

	/**
	 * The file CRCs.
	 */
	private final int[] crcs;

	/**
	 * Creates the CrcList.
	 *
	 * @param crcs The file CRCs. Further changes to this array will not affect this list.
	 */
	public CrcList(int[] crcs) {
		this.crcs = crcs.clone();
	}

	/**
	 * Gets the CRC of the specified file.
	 * 
	 * @param file The file.
	 * @return The CRC.
	 */
	public int getCrc(int file) {
		Preconditions.checkElementIndex(file, crcs.length, "CRC file id out of bounds.");
		return crcs[file];
	}

	/**
	 * Gets the array of file CRCs. Changes to this array will not affect this list.
	 * 
	 * @return The CRCs.
	 */
	public int[] getCrcs() {
		return crcs.clone();
	}

	/**
	 * Gets the amount of file CRCs.
	 * 
	 * @return The amount.
	 */
	public int size() {
		return crcs.length;
	}

}

File: legacy/src/main/rs/emulate/legacy/version/CrcList.kt
package rs.emulate.legacy.version

/**
 * A list of file CRCs, for a particular ArchiveEntry (such as models).
 *
 * @param type The [VersionEntryType] of the CrcList.
 * @param crcs The file CRCs. Further changes to this array will not affect this list.
 */
class CrcList(val type: VersionEntryType, crcs: IntArray) {

    val crcs: IntArray = crcs.clone()
        get() = field.clone()

    /**
     * Gets the CRC of the specified file.
     */
    operator fun get(file: Int): Int {
        require(file < crcs.size) { "CRC file id out of bounds." }
        return crcs[file]
    }

    /**
     * Gets the amount of file CRCs.
     */
    val size: Int
        get() = crcs.size

    companion object {

        /**
         * The name suffix of CrcList entries.
         */
        internal const val ENTRY_NAME_SUFFIX = "_crc"
    }

}
